dattri/algorithm/dvemb.py

```diff
@@ -0,0 +1,588 @@
+                                "none"(default),
+                                "kronecker"(same with paper),
+                                or "elementwise"(more memory-efficient).
+
+        Raises:
+            ValueError: If an unknown factorization type is provided.
+        """
+        self.model = model
+        self.device = device
+        self.criterion = loss_func
+        self.projector = None
+        self.use_factorization = factorization_type != "none"
+        self.factorization_type = factorization_type
+        self.projection_dim = proj_dim
+
+        if not self.use_factorization:
+
+            @flatten_func(self.model)
+            def vmap_loss_fn(
+                params: Tensor,
+                data_tensors: Tuple[Tensor, Tensor],
+            ) -> Tensor:
+                inputs = data_tensors[0].unsqueeze(0)
+                targets = data_tensors[1].unsqueeze(0)
+                outputs = torch.func.functional_call(self.model, params, inputs)
+                return self.criterion(outputs, targets)
+
+            self.per_sample_loss_fn = vmap_loss_fn
+
+        if self.use_factorization:
+            self.cached_factors: Dict[int, List[List[Dict[str, Tensor]]]] = {}
+            self._linear_layers = list(self._iter_linear_layers(self.model))
+            if proj_dim is None:
+                self._params_dim = sum(
+                    p.numel()
+                    for layer in self._linear_layers
+                    for p in layer.parameters()
+                )
+            else:
+                if self.factorization_type == "kronecker":
+                    self.projection_dim = int(
+                        math.sqrt(proj_dim / len(self._linear_layers)),
+                    )
+                    self._params_dim = (
+                        len(self._linear_layers)
+                        * self.projection_dim
+                        * self.projection_dim
+                    )
+                elif self.factorization_type == "elementwise":
+                    self.projection_dim = int(proj_dim / len(self._linear_layers))
+                    self._params_dim = len(self._linear_layers) * self.projection_dim
+                else:
+                    msg = f"Unknown factorization type: {self.factorization_type}"
+                    raise ValueError(msg)
+                # projectors for each gradient factor (d1 x k) and (d2 x k)
+                self.random_projectors = []
+                for layer in self._linear_layers:
+                    output_dim = layer.weight.shape[0]
+                    input_dim = layer.weight.shape[1]
+                    project_output = self._generate_projector(
+                        output_dim,
+                        self.projection_dim,
+                    )
+                    project_input = self._generate_projector(
+                        input_dim,
+                        self.projection_dim,
+                    )
```

Comment:
Let's use the projector once https://github.com/TRAIS-Lab/dattri/pull/203 is merged.